test/test_um2netcdf.py

```diff
@@ -5,7 +5,209 @@
+
+
+@pytest.fixture
+def mule_vars(z_sea_rho_data, z_sea_theta_data):
+    """Simulate mule variables from aiihca.paa1jan.subset data."""
+    d_lat = 1.25  # spacing manually copied from aiihca.paa1jan.subset file
+    d_lon = 1.875
+    return um2nc.MuleVars(um2nc.GRID_NEW_DYNAMICS, d_lat, d_lon, z_sea_rho_data, z_sea_theta_data)
+
+
+@pytest.fixture
+def air_temp_cube():
+    # copied from aiihca.paa1jan.subset file
+    return DummyCube(30204, "air_temperature")
+
+
+@pytest.fixture
+def precipitation_flux_cube():
+    # copied from aiihca.paa1jan.subset file
+    return DummyCube(5216, "precipitation_flux")
+
+
+@pytest.fixture
+def std_args():
+    # TODO: make args namedtuple?
+    args = mock.Mock()
+    args.nomask = False
+    args.nohist = False
+    args.nckind = 3
+    args.include_list = None
+    args.exclude_list = None
+    args.simple = False
+    args.verbose = False
+    return args
+
+
+def test_process_without_masking(air_temp_cube, precipitation_flux_cube, mule_vars, std_args):
+    """Attempts end-to-end test of process(), ignoring cubes requiring masking."""
+
+    # FIXME: this convoluted setup is a big code stench
+    #        use this to gradually refactor process()
+    #        add naming vars to the dummy cubes
+    with mock.patch("mule.load_umfile"):  # ignore m_load_umfile as process_mule_vars is mocked
+        with mock.patch("umpost.um2netcdf.process_mule_vars") as m_mule_vars:
+            m_mule_vars.return_value = mule_vars
+
+            with mock.patch("iris.load") as m_iris_load:
+                cubes = [air_temp_cube, precipitation_flux_cube]
+
+                for c in cubes:
+                    c.attributes = {um2nc.STASH: DummyStash(*split_item_code(c.item_code))}
+                    c.cell_methods = []
+
+                    # TODO: replace with DummyCubeWithCoords ()?
+                    #       or add coord fixtures?
+                    c.coord["latitude"] = 0.0  # FIXME
+                    c.coord["longitude"] = 0.0  # FIXME
+
+                m_iris_load.return_value = cubes
+
+                with mock.patch("iris.fileformats.netcdf.Saver") as m_saver:  # prevent I/O
+                    # mock `sman` var to prevent I/O
+                    m_sman = mock.Mock()
+                    m_saver().__enter__.return_value = m_sman
+
+                    # TODO: fix lat/lon & levels requires c.coord attribute
+                    #       use fixtures to add attrs & remove the patches?
+                    with mock.patch("umpost.um2netcdf.fix_latlon_coord") as m_coord:
+                        with mock.patch("umpost.um2netcdf.fix_level_coord") as m_level:
+                            with mock.patch("umpost.um2netcdf.apply_mask") as m_apply_mask:
+                                with mock.patch("umpost.um2netcdf.cubewrite") as m_cubewrite:
+                                    infile = "/tmp/fake_input_fields_file"
+                                    outfile = "/tmp/fake_input_fields_file.nc"
```

Comment:
Can the tempfile module be used here instead of direct reference to /tmp?

Reply:
It can & `pytest`'s `tmp_path` fixture has similar functionality. The downside is that both touch the filesystem, which I'm attempting to avoid with the unit tests. I'm working to a 3 level testing strategy:
1. `Unit:` external dependencies are mocked, tests are fast quick feedback during dev
2. `Integration:` command line run(s) against a few data files (slower)
3. `Full:` process a collection of multiple real data files

I could replace the `/tmp/fake...` paths with nonsense paths that don't exist or will fail due to permissions should the mocks fail for some reason.

Reply:
Ah, right. So if I'm reading this correctly, the `mock.patch` calls above have disabled all of the I/O functionality, so you're just sending in fake filenames so there's something for `process()` to absorb.

I like the idea of replacing them with complete garbage though; as you mention, it provides a level of cover against mock failure (in particular, you won't try to play with a system's `/tmp` directory should the mocks fail out).

Reply:
> Ah, right. So if I'm reading this correctly, the `mock.patch` calls above have disabled all of the I/O functionality

Yep, code like the `mock.patch("mule.load_umfile")` call dynamically replaces the real `mule.load_umfile()` with a generic mock object that mostly just records calls & attr lookups made to it. The I/O func calls are redirected to the mock & the `process()` code continues obliviously :-)

It's not ideal, but the project imposes a few constraints where mocking helps.

> so you're just sending in fake filenames so there's something for `process()` to absorb. (Update: I missed the mock to `Saver` further up, c.f. L150-155).

Yep, the mock just "absorbs" the call, records the details & returns control.

A future step is refactoring the input I/O calls to a small func & moving `process()`'s core to func that **only** takes data args (thus avoids filesystem I/O). Once the core takes data only args, some mocking can be discarded as the tests should just provide raw data to process. Getting there requires a few evolutionary refactoring steps.

> I like the idea of replacing them with complete garbage though; as you mention, it provides a level of cover against mock failure (in particular, you won't try to play with a system's `/tmp` directory should the mocks fail out).

This is fixed in https://github.com/ACCESS-NRI/um2nc-standalone/pull/58/commits/f2c628c610629a2f459ea11ea1a64f8ef4fa67bb.